### ISSUE TITLE
AcquireTokenPopup for someone other than logged on user

### ISSUE THREAD

**Issue by Ivan-L** [AzureAD/microsoft-authentication-library-for-js#1293]
## Library
- [X] `msal@1.x.x` or `@azure/msal@1.x.x`

## Description
This question relates to `msal@1.2.x` even though we have SPA which uses `@azure/msal-angular@0.1.4-beta.2` and which we're trying to upgrade to `@azure/msal-angular@1.0.0-beta.2` (therefore bringing in the `msal@1.2.x` peer dependency).

Previously when we were using `@azure/msal-angular@0.1.4-beta.2`, we took a dependency of `@azure/msal-angular/msal@0.2.4`. Our use case worked with `@azure/msal-angular/msal@0.2.4` and it no longer works with `msal@1.2.x`.

Our use case is as follows. There is a main logged on user in our app who performs various activities in the app and at some point needs to perform a privileged operation that requires elevated permissions. The user does not have those permissions and needs to call over a manager to type in his or her credentials. Barring the inherent risk of the manager typing his or her credentials on a user workstation (which we plan on curing by enabling 2FA for the manager), this scenario worked fine when we used `acquireTokenPopup` in `@azure/msal-angular/msal@0.2.4` using the following code:
```
const user = this.authService.getUser();
return from(this.authService.acquireTokenPopup(
  ['api://[our_api_uri]/user_impersonation`],
  'https://login.microsoftonline.com/[tenantId],
   user,
   'prompt=login&forceRefresh=true'
));
```
The manager would sign in using his or her own credentials and this would give us an access token which we then used to call our API to execute the privileged operation as the manager. This did not affect the currently logged on user and after the privileged operation was executed, the user would continue performing functions in the app as himself/herself. The manager therefore provided an access token which was used one time, just for the privileged operation.

After upgrading to `msal@1.2.2-beta.0` which is a peer dependency of `@azure/msal-angular@1.0.0-beta.2`, we needed to change the above code to the following:
```
const account = this.authService.getAccount();
return from(this.authService.acquireTokenPopup({
  scopes: [environment.api.audience + '/user_impersonation'],
  account: account,
  prompt: 'login',
  forceRefresh: true
})).pipe(
  map(authResponse => authResponse.accessToken)
);
```

Seems like a simple enough upgrade, right? Unfortunately not. The `acquireTokenPopup` returns a `null` `accessToken` in the updated code. After digging around and debugging the code inside `UserAgentApplication`, I found that even though Azure AD was returning an `accessToken` for the correct account (the manager) in the URL hash, it was being discarded in the [following code](https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/1212c640aa619718328a5f6931f14ecf27ea0b36/lib/msal-core/src/UserAgentApplication.ts#L1493-L1504) within the `saveTokenFromHash` method in `UserAgentApplication`:
```
// Process access_token
if (hashParams.hasOwnProperty(ServerHashParamKeys.ACCESS_TOKEN)) {
    this.logger.info("Fragment has access token");

    // retrieve the id_token from response if present
    if (hashParams.hasOwnProperty(ServerHashParamKeys.ID_TOKEN)) {
        idTokenObj = new IdToken(hashParams[ServerHashParamKeys.ID_TOKEN]);
        response.idToken = idTokenObj;
        response.idTokenClaims = idTokenObj.claims;
    } else {
        idTokenObj = new IdToken(this.cacheStorage.getItem(PersistentCacheKeys.IDTOKEN));
        response = ResponseUtils.setResponseIdToken(response, idTokenObj);
    }
```
So in the above, `hashParams[ServerHashParamKeys.ACCESS_TOKEN]` is not used at all in the resultant response. What we get back is actually a response with the `idToken` set for the currently logged on user instead of a response with an `accessToken` for the manager.

The questions are therefore:

1. Is this a bug, and if not,
2. Is there a way that we can get an access token for the manager using `msal@1.2.x` without affecting the currently logged on user and also without affecting any associated session or local storage for the user?

**Comment by jasonnutter** [comment]
@Ivan-L Thanks for the detailed investigation! This definitely looks like a bug. Have you had a chance to figure out what a potential fix might be? 

Regardless, we'll investigate and follow up, thanks!

**Comment by Ivan-L** [comment]
@jasonnutter Apologies for the late response, I have been under pressure to meet certain deadlines.

I did some more digging and narrowed down the issue even further by debugging both the old version and new version of `UserAgentApplication`.

In the old version 0.2.4 of the `processCallback` method of `UserAgentApplication`, the access token in our case [is obtained](https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/e9ed1408464c38d0cf6ccd8fb0ca85965c5218b8/lib/msal-core/src/UserAgentApplication.ts#L310) regardless of what the `saveTokenFromHash` does just before that. Then, the token received callback [is called](https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/v0.2.4/lib/msal-core/src/UserAgentApplication.ts#L327) with the correct token. This is why our scenario worked in the past.

In the newer versions, the `AuthResponse` class was introduced and reliance is placed on `saveTokenFromHash` populating the `AuthResponse` correctly. The only "post processing" that `processCallback` does after `saveTokenFromHash` is called, is that [the token type is set](https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/1212c640aa619718328a5f6931f14ecf27ea0b36/lib/msal-core/src/UserAgentApplication.ts#L997) in the response. The `processCallback` method no longer tries to get the access token from the hash anymore.

So my suggestion would be to update the `saveTokenFromHash` method, specifically the [block of code](https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/1212c640aa619718328a5f6931f14ecf27ea0b36/lib/msal-core/src/UserAgentApplication.ts#L1493-L1504) that I pasted previously, to not only try to get the ID token from the hash but also get the access token from the hash as well.
The following code:
```ts
// Process access_token
if (hashParams.hasOwnProperty(ServerHashParamKeys.ACCESS_TOKEN)) {
    this.logger.info("Fragment has access token");

    // retrieve the id_token from response if present
    if (hashParams.hasOwnProperty(ServerHashParamKeys.ID_TOKEN)) {
        idTokenObj = new IdToken(hashParams[ServerHashParamKeys.ID_TOKEN]);
        response.idToken = idTokenObj;
        response.idTokenClaims = idTokenObj.claims;
    } else {
        idTokenObj = new IdToken(this.cacheStorage.getItem(PersistentCacheKeys.IDTOKEN));
        response = ResponseUtils.setResponseIdToken(response, idTokenObj);
    }
    ...
```
Could change to:
```ts
// Process access_token
if (hashParams.hasOwnProperty(ServerHashParamKeys.ACCESS_TOKEN)) {
    this.logger.info("Fragment has access token");

    /***** START NEW CODE ******/
    // retrieve the access_token from hashParams since we already know it's there
    response.accessToken = hashParams[ServerHashParamKeys.ACCESS_TOKEN];
    /***** END NEW CODE ******/

    // retrieve the id_token from response if present
    if (hashParams.hasOwnProperty(ServerHashParamKeys.ID_TOKEN)) {
        idTokenObj = new IdToken(hashParams[ServerHashParamKeys.ID_TOKEN]);
        response.idToken = idTokenObj;
        response.idTokenClaims = idTokenObj.claims;
    } else {
        idTokenObj = new IdToken(this.cacheStorage.getItem(PersistentCacheKeys.IDTOKEN));
        response = ResponseUtils.setResponseIdToken(response, idTokenObj);
    }
    ...
```

After the above change, the response should contain the access token as well and it would no longer be discarded.

As a separate issue, I do not think that the handling of the ID token is quite correct in the existing code because the code falls back to using the existing ID token in the cache if the hash does not contain one. In our case we request an access token for a user who is different to the currently logged on user through the `acquireTokenPopup` method, yet we get back an `AuthResponse` response with the ID token set as the currently logged on user. But as mentioned, this is a separate issue that should be debated.

**Comment by jasonnutter** [comment]
@Ivan-L Thanks! This is very helpful. I'll put up a PR, and we'll get it out in the next version / beta.

**Comment by jasonnutter** [comment]
@Ivan-L Question: in your scenario, does this log statement get reached? https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/1212c640aa619718328a5f6931f14ecf27ea0b36/lib/msal-core/src/UserAgentApplication.ts#L1541-L1542

If not, which path does it take?

**Comment by Ivan-L** [comment]
@jasonnutter No, the log statement does not get reached.

Neither
```
if (!StringUtils.isEmpty(cachedAccount)) {
```
nor
```
else if (!StringUtils.isEmpty(this.cacheStorage.getItem(acquireTokenAccountKey_noaccount))) {
```
evaluate to `true` and the entire [block of code](https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/1212c640aa619718328a5f6931f14ecf27ea0b36/lib/msal-core/src/UserAgentApplication.ts#L1534-L1547) is skipped in our scenario.

I have not tested the suggestion in our app. I can do so, but it may take me some time to clone this repo and figure out the build and symlink-ing needed to test out the change. I can give it a shot in the evening.

**Comment by jasonnutter** [comment]
@Ivan-L I see, thanks for the follow up. Don't worry about testing this out yet, I'll chat with the team today and if we like your proposed fix, we can likely release a beta version with a fix that you can try.

**Comment by jasonnutter** [comment]
The fix for this is available in `msal@1.2.2-beta.2`, please try that version and confirm that it has been fixed, thanks!

**Comment by Ivan-L** [comment]
The issue has been fixed with `msal@1.2.2-beta.2`, thanks @jasonnutter!
